### PR TITLE
Adjust PathLen checks for pre-issued precerts

### PR DESCRIPTION
**Pull request by daviddrysdale**

### CODE REVIEW

File: trillian/testdata/Makefile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 all: pubkeys ca leaves
 
 # The following private keys are never regenerated.
-SERVER_PRIVKEYS=ct-http-server.privkey.pem log-rpc-server.privkey.pem map-rpc-server.privkey.pem
+SERVER_PRIVKEYS=ct-http-server.privkey.pem log-rpc-server.privkey.pem
```
Comment (Martin2112) [inline trillian/testdata/Makefile]: Seems unrelated?

File: trillian/testdata/int-ca.cert [hunk 2]
```diff
@@ -1,14 +1,14 @@
 -----BEGIN CERTIFICATE-----
-MIICGzCCAcKgAwIBAgIEQkJCQjAKBggqhkjOPQQDAjBxMQswCQYDVQQGEwJHQjEP
+MIICHTCCAcKgAwIBAgIEQkJCQjAKBggqhkjOPQQDAjBxMQswCQYDVQQGEwJHQjEP
```
Comment (Martin2112) [inline trillian/testdata/int-ca.cert]: This change updates a lot of test files. It was hard to tel if we're testing all the old cases + new ones?

File: trillian/ctfe/cert_checker.go [hunk 3]
```diff
@@ -118,25 +118,44 @@ func ValidateChain(rawChain [][]byte, validationOpts CertValidationOpts) ([]*x50
...
 			if eku == x509.ExtKeyUsageCertificateTransparency {
 				originalEKUs = chain[1].ExtKeyUsage
 				chain[1].ExtKeyUsage = append(chain[1].ExtKeyUsage[:i], chain[1].ExtKeyUsage[i+1:]...)
+				havePreissuer = true
 				break
 			}
 		}
 	}
 
+	if havePreissuer {
```
Comment (Martin2112) [inline trillian/ctfe/cert_checker.go]: What would the cost be of making a new modified chain? Then it wouldn't have to be undone.

File: trillian/testdata/Makefile (hunk 1 above)
Comment (daviddrysdale) [reply trillian/testdata/Makefile]: Yeah, it is a bit -- this is (I think) the first time any of the testdata has been rebuilt since the repo split, and the `Makefile` didn't get fully updated at the time.  I'll pull it out into a separate commit

File: trillian/testdata/int-ca.cert (hunk 2 above)
Comment (daviddrysdale) [reply trillian/testdata/int-ca.cert]: Yeah, all the test data gets regenerated but it's mechanical from the `Makefile`.

(It would probably be / have been better to have the synthetic certs entirely generated in code rather than read from a known `../testdata` directory, but that's a separate question/PR).

File: trillian/ctfe/cert_checker.go (hunk 3 above)
Comment (daviddrysdale) [reply trillian/ctfe/cert_checker.go]: The runtime cost isn't a worry, it's the swapping things around that would be fiddly and error-prone because everything uses pointers to the same set of `x509.Certificates`.  The call to `chain[0].Verify()` generates the returned chain, but that in turn relies on the intermediate pool that's populated from the chain.  If we had two copies of the chain (modified and unmodified) we'd have to do index mapping from one to the other to rebuild the analogous chain of unmodified certs.

File: trillian/testdata/int-ca.cert (hunk 2 above)
Comment (daviddrysdale) [reply trillian/testdata/int-ca.cert]: (After in-person discussion I now understand what you're asking)

This change does alter the integration test from testing (chains with no intermediate path-len constraints) to testing (chains with intermediate path-len=0) -- but I don't think that losing the former scenario is a significant loss.

Comment (Martin2112) [reply trillian/testdata/int-ca.cert]: Ack.
